src/box/iproto.cc

```diff
@@ -2225,6 +2248,114 @@ iproto_is_connection_shutdown_ready(struct iproto_connection *con)
 	       (!con->graceful_shutdown && con->socket_shutdown);
 }
 
+/**
+ * Start shutdown process on given @a connection.
+ */
+static inline void
+iproto_start_connection_shutdown(struct iproto_connection *con)
+{
+	assert(con->iproto_thread->is_shutdown_active);
+	ev_async_start(con->loop, &con->after_shutdown);
+	cmsg_init(&con->shutdown_msg, con->iproto_thread->shutdown_route);
+	cpipe_push(&con->iproto_thread->tx_pipe, &con->shutdown_msg);
+	cpipe_flush_input(&con->iproto_thread->tx_pipe);
+}
+
+/**
+ * Choose the way for shutdown process depending on whether the connection
+ * supports graceful shutdown shutdown or not.
+ */
+static void
+tx_start_shutdown(struct cmsg *m)
+{
+	struct iproto_connection *con =
+		container_of(m, struct iproto_connection, shutdown_msg);
+	struct cpipe *net_pipe = &con->iproto_thread->net_pipe;
+	if (con->session && con->session->graceful_shutdown) {
+		cpipe_push(net_pipe, &con->graceful_shutdown_msg);
+	} else {
+		cmsg_init(&con->not_graceful_shutdown_msg,
+			  con->iproto_thread->not_graceful_shutdown_route);
+		cpipe_push(net_pipe, &con->not_graceful_shutdown_msg);
+	}
+	cpipe_flush_input(net_pipe);
+}
+
+/**
+ * Send IPROTO_SHUTDOWN to a remote client.
+ */
+static void
+net_send_shutdown(struct cmsg *m)
+{
+	struct iproto_connection *con =
+		container_of(m, struct iproto_connection, shutdown_msg);
+	assert(con->iproto_thread->is_shutdown_active);
+	con->graceful_shutdown = true;
+	iproto_reply_shutdown(con->tx.p_obuf, ::schema_version);
+	if (!con->tx.is_push_sent)
+		tx_begin_push(con);
+	else
+		con->tx.is_push_pending = true;
+	if (iproto_is_connection_shutdown_ready(con))
+		ev_async_send(con->loop, &con->after_shutdown);
+}
+
+/**
+ * Read input socket of connection until iproto packet boundary or until
+ * msg_max_limit is reached and shutdown(con, SHUT_RD).
+ */
+static void
+net_shutdown_socket(struct cmsg *m)
+{
+	struct iproto_connection *con =
+		container_of(m, struct iproto_connection, shutdown_msg);
+	assert(con->iproto_thread->is_shutdown_active);
+	con->graceful_shutdown = false;
+	int fd = con->input.fd;
+	if (fd < 0)
+		return;
+	try {
+		iproto_resume(con->iproto_thread);
+		/* Read input while input exists and msg_max_limit isn't
+		 * reached.
```

Comment:
But more importantly, I don't understand why do you even need to read the pending messages. It does not give any guarantees, because, for example, the message could just not arrive in time, but after the shutdown. If a client didn't receive a response for something, and the request was not already started, the guarantee count is 0.